examples/generative/finetune_stable_diffusion.py

```diff
@@ -0,0 +1,507 @@
+    dataset = dataset.map(process_image, num_parallel_calls=AUTO).batch(batch_size)
+    dataset = dataset.map(apply_augmentation, num_parallel_calls=AUTO)
+    dataset = dataset.map(run_text_encoder, num_parallel_calls=AUTO)
+    dataset = dataset.map(prepare_dict, num_parallel_calls=AUTO)
+    return dataset.prefetch(AUTO)
+
+
+"""
+The baseline Stable Diffusion model was trained using images of 512x512 resolution. It's
+unlikely for a model that's trained using higher resolution images to transfer well to
+lower resolution images. However, the current model will lead to OOM if we keep the
+resolution to 512x512 (without enabling mixed-precision). Therefore, in the interest of
+interactive demonstration, we kept the input resolution to 256x256.
+"""
+
+# Prepare the dataset.
+training_dataset = prepare_dataset(
+    np.array(data_frame["image_path"]), tokenized_texts, batch_size=4
+)
+
+# Take a sample batch and investigate.
+sample_batch = next(iter(training_dataset))
+
+for k in sample_batch:
+    print(k, sample_batch[k].shape)
+
+"""
+We can also take a look at the training images and their corresponding captions.
+"""
+
+plt.figure(figsize=(20, 10))
+
+for i in range(3):
+    ax = plt.subplot(1, 4, i + 1)
+    plt.imshow((sample_batch["images"][i] + 1) / 2)
+
+    text = tokenizer.decode(sample_batch["tokens"][i].numpy().squeeze())
+    text = text.replace("<|startoftext|>", "")
+    text = text.replace("<|endoftext|>", "")
+    plt.title(text)
+
+    plt.axis("off")
+
+"""
+## A trainer class for the fine-tuning loop
+"""
+
+
+class Trainer(tf.keras.Model):
+    # Reference:
+    # https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py
+
+    def __init__(
+        self,
+        diffusion_model,
+        vae,
+        noise_scheduler,
+        use_mixed_precision=False,
+        max_grad_norm=1.0,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        self.diffusion_model = diffusion_model
+        self.vae = vae
+        self.noise_scheduler = noise_scheduler
+        self.max_grad_norm = max_grad_norm
+
+        self.use_mixed_precision = use_mixed_precision
+        self.vae.trainable = False
+
+    def train_step(self, inputs):
+        images = inputs["images"]
+        encoded_text = inputs["encoded_text"]
+        batch_size = tf.shape(images)[0]
+
+        with tf.GradientTape() as tape:
+            # Project image into the latent space and sample from it.
+            latents = self.sample_from_encoder_outputs(self.vae(images, training=False))
+            # Know more about the magic number here:
+            # https://keras.io/examples/generative/fine_tune_via_textual_inversion/
+            latents = latents * 0.18215
+
+            # Sample noise that we'll add to the latents.
+            noise = tf.random.normal(tf.shape(latents))
+
+            # Sample a random timestep for each image.
+            timesteps = tnp.random.randint(
```

Comment:
nit: timestep -> sampled_t?

Reply:
> why not just define a macro BATCH_SIZE=4

That impacts the dynamic behavior of the code IMO. In the current case, users can just pass on a single batch size and let the code take care of the rest. Also, increasing batch size, in this case, might easily lead to OOM, so that option is a bit restricted. 

> nit: timestep -> sampled_t?

Not consistent with the rest of the guides we have and the existing codebases on Stable Diffusion.